A Cdi Portable Extension Example adding a custom @FactModel based on Drools

marcobattaglia opened the pull request:
This is a Quick-Start showing how to augment javaEE6 adding a simple Drools based extension to decouple business logic and presentation layer. It uses Arquillian Drone to test.

Commit:
newCdiExampleProject

Added explicit check on @FactModel annotation presence
before augmenting class.
Only for educational purpose.

Removed duplicated file arquillian.xml
Commented arquillian.xml optional tags
Removed unused void constructor in org/jboss/as/quickstarts/cdi/drools/extension/FactMethod.java
Update uncorrect english grammar syntax in README.md ;-)

Slimming POM.xml and test class

Slimming pom.xml and test class

Blank spaces are removed from README.md

Removing return carridge for a better html view

Updated the new container configuration section

Added some comments to classes, methods, rule.
Extended description of README.md explaining better Drools and added references to more significant classes
Re-organized rules to having an example closer to real sistuations.
Added three tests to check results of adding and removing basket items.
Now there are three rules: Normal, Silver and Gold that defines three levels of discount.
It's a little bit complex, but it is more realistic and readable than the previous version.

Capitalized s of Selenium

Organized in different paragraphs

Deleted first row

modified READ.ME: The target product is BRMS
modified LineItemInjectionTarget .PMuir has suggested to augment a factmodel inside produce(CreationalContext<X> ctx) method. In produceMethod rsession is still null. I've moved this code inside postconstruct method. This guarantees a mapping one to one from factModel and its fact rapresentation in ksession. Deleted "*" char in wrong position.
modified DroolsExtension comments
modified RuleSessionManager: added correct access modfier, added comments
modified Changed the  name of FactMethod to FactMethodInterceptor
modified RuleSessionManager: used a generated serialVersionNumber, added comments
modified Manager set as private some fields. Simplified the production and injection of baskeItem.
modified OrderBuilder: comments added.
modified files of resources to be loaded into arquillian .war a  generated ar file

Simplified the expression "behind the JSF page, is made up of..."
Added emphasis on Test section
Corrected /s/taht/that

Added references to drools bom

Added bom references to drools

Added jboss-javaee-6.0-with-drools bom reference

Code formatted and cleaned up

Example part is been indented

LightGuard commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,125 @@
+JSF view is managed by a portable extension as a Fact registered in a ksession, so that the rule engine acts as 
+the decoupled logic system behind the presentation layer.
+
+In this project developer could manage the ksession and its scope using a producer.
```
Missing something in this sentence. "a" developer perhaps

LightGuard commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,316 @@
+			<artifactId>drools-compiler</artifactId>
+			<version>[5.3.0.Final]</version>
+		</dependency>
+
```
If we don't already have a BOM which includes Drools, we need one. Then these can go away.

LightGuard commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,316 @@
+			<groupId>org.apache.deltaspike.core</groupId>
+			<artifactId>deltaspike-core-api</artifactId>
+			<version>LATEST</version>
+		</dependency>
```
This should be using a version defined in a BOM

LightGuard commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,316 @@
+			<artifactId>deltaspike-core-impl</artifactId>
+			<version>LATEST</version>
+			<scope>runtime</scope>
+		</dependency>
```
Same as above

LightGuard commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,316 @@
+			<groupId>org.jboss.weld</groupId>
+			<artifactId>weld-core-test-arquillian</artifactId>
+			<version>1.1.10.Final</version>
+			<scope>test</scope>
```
Is this needed if you're going to be deploying to the server?

marcobattaglia replied:
Without weld dependency test fails. Perhaps there is an error in test file. I'm checking. Thanx

LightGuard replied:
That's very odd. Looking at the test classes, if you're running with the arq-jbossas-managed or arq-jbossas-remote profiles things should be fine.

marcobattaglia replied:
My test is annotated as 
@Test
@RunAsClient
I've added weld-core-test-arquillian as documented here ("you have to specify the dependency on container adapter") http://arquillian.org/guides/functional_testing_using_drone/#set_up_drone. I can try a workaround.

LightGuard replied:
Are you wanting to test in AS / EAP or a stand-alone weld container (Weld
SE)?

On Fri, Feb 15, 2013 at 9:52 AM, marcobattaglia notifications@github.comwrote:

> In cdi-portable-extension-drools/pom.xml:
> 
> > -                   <groupId>org.mortbay.jetty</groupId>
> > -                   <artifactId>servlet-api-2.5</artifactId>
> > -               </exclusion>
> > -           </exclusions>
> > -       </dependency>
> > -       <dependency>
> > -           <groupId>org.slf4j</groupId>
> > -           <artifactId>slf4j-simple</artifactId>
> > -           <version>1.6.4</version>
> > -           <scope>test</scope>
> > -       </dependency>
> > -       <dependency>
> > -           <groupId>org.jboss.weld</groupId>
> > -           <artifactId>weld-core-test-arquillian</artifactId>
> > -           <version>1.1.10.Final</version>
> > -           <scope>test</scope>
> 
> My test is annotated as
> @Test https://github.com/Test
> @RunAsClient
> I've added weld-core-test-arquillian as documented here ("you have to
> specify the dependency on container adapter")
> http://arquillian.org/guides/functional_testing_using_drone/#set_up_drone.
> I can try a workaround.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jboss-jdf/jboss-as-quickstart/pull/439/files#r3031472.

## 

Jason Porter
http://en.gravatar.com/lightguardjp

marcobattaglia replied:
I'm testing in AS/EAP. The strange thing is that it runs with <scope>provided</scope> but not with <scope>runtime</scope>. It seems that this dependency simple activates the reference to the external jboss container, infact without it my jboss doesn't start (using arq-jbossas-managed) or my app isn't deployed (using arq-jbossas-remote). The question is: weld-core-test-arquillian is only an adapter that interacts with an external container or is it the whole container?
Thanx

LightGuard replied:
This seems very odd to me, but I'm not familiar with Drone

aslakknutsen replied:
This dependency should not be there. This is the Weld Core Test Suite, not the container. If it's needed it's only because you've got 'something' else wrong and this artifacts transitive deps 'fix' it.

marcobattaglia replied:
OK. I check to resolve these dependencies in a better way and I'll commit. Thanx

LightGuard commented on cdi-portable-extension-drools/resources/arquillian.xml:
```diff
@@ -0,0 +1,31 @@
+    <!-- Uncomment to have test archives exported to the file system for inspection -->
+    <engine>
+        <property name="deploymentExportPath">target/</property>
+    </engine>
```
This should be commented out by default

LightGuard commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/FactMethod.java:
```diff
@@ -0,0 +1,41 @@
+
+	public FactMethod() {
+		// TODO Auto-generated constructor stub
+	}
```
The ctor could be removed

LightGuard commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/FactMethod.java:
```diff
@@ -0,0 +1,41 @@
+	}
+    
+	@AroundInvoke
+	
```
Remove the space

LightGuard commented on cdi-portable-extension-drools/resources/arquillian.xml:
```diff
@@ -0,0 +1,31 @@
+        <!--<property name="jbossHome">/path/to/jboss/as</property> -->
+        <!--</configuration> -->
+    </container>
+</arquillian>
```
After going through all of the files, why is this one here? The one in src/test/resources should be the only one needed.

aslakknutsen commented on cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:
```diff
@@ -0,0 +1,89 @@
+    }
+   
+	@Drone
+    DefaultSelenium browser; 
```
Any good reason to use DefaultSelenium instead of the newer WebDriver API?

aslakknutsen commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,316 @@
+					<artifactId>servlet-api-2.5</artifactId>
+				</exclusion>
+			</exclusions>
+		</dependency>
```
Replace all arquillian-drone\* and org.selenium\* with the drone-webdriver-depchain. 

org.jboss.arquillian.extension:arquillian-drone-webdriver-depchain

aslakknutsen commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,125 @@
+
+Results :
+
+Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```
Missing spacing to include full output in markdown 'source block'

aslakknutsen commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,125 @@
+22:42:37.962 INFO - Got result: OK,true on session 2f1327a4691e40af9af285ee7bb0461e
+....
+22:42:38.254 INFO - Got result: OK on session 2f1327a4691e40af9af285ee7bb0461e
+Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 27.014 sec
```
Missing spacing to include full output in markdown 'source block'

aslakknutsen commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,316 @@
+			</dependency>
+			<dependency>
+				<groupId>org.jboss.arquillian</groupId>
+				<artifactId>arquillian-bom</artifactId>
```
Arquillian-bom should be included and managed by jboss-javaee-6.0-with-tools already, so this is not needed.

aslakknutsen commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,316 @@
+			</dependency>
+			<dependency>
+				<groupId>org.jboss.arquillian.extension</groupId>
+				<artifactId>arquillian-drone-bom</artifactId>
```
Arquillian-drone-bom should be included and managed by jboss-javaee-6.0-with-tools already, so this is not needed.

aslakknutsen commented on cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:
```diff
@@ -0,0 +1,89 @@
+                        "org.drools:knowledge-api",
+                        "org.drools:drools-core",
+                        "org.drools:drools-compiler",
+                        "org.seleniumhq.selenium:selenium-java").resolveAsFiles()
```
seleniumhq should not be needed in the deployment, it's a client side lib

aslakknutsen commented on cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:
```diff
@@ -0,0 +1,89 @@
+	
+	
+   @Test
+   @RunAsClient
```
@RunAsClient not needed here since @Deployment.testable = false . RunAsClient is the only available run mode in this case so its inferred.

aslakknutsen commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/FactMethod.java:
```diff
@@ -0,0 +1,36 @@
+public class FactMethod {
+	@Inject  @RuleSession RuleSessionManager rsession;
+        
+        @AroundInvoke
```
formatting is a bit off here

aslakknutsen commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/model/RuleSessionProducer.java:
```diff
@@ -0,0 +1,34 @@
+public class RuleSessionProducer {
+
+	@Produces @RuleSession @ConversationScoped 
+		 RuleSessionManager getRuleSessionManager(){
```
formatting is a bit off here

aslakknutsen commented on cdi-portable-extension-drools/src/main/resources/SimpleRule.drl:
```diff
@@ -0,0 +1,16 @@
+    when
+       $lineItem:LineItem(discount == 0.0, value > 0 );					         
+	then
+          modify($lineItem){
```
formatting is a bit off here

aslakknutsen commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/LineItemInjectionTarget.java:
```diff
@@ -0,0 +1,85 @@
+	    }
+
+	    @Override
+	    public Set<InjectionPoint> getInjectionPoints() {
```
this class has a different indent level then the others

sgilda commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,184 @@
+ProcessAnnotatedType and interceptor bindings from CDI, and AnnotatedTypeBuilder from DeltaSpike
+to demonstrate how you can augment JavaEE6 with custom extension. A model of a
+JSF view is managed by a portable extension as a Fact registered in a ksession, so that the rule engine acts as 
+the decoupled logic system behind the presentation layer.
```
Hi Marco, would you mind removing the carriage-returns from these and other lines in this file. We only use line feeds for paragraphs because the HTML can render strangely when you use them.

sgilda commented on cdi-portable-extension-drools/src/test/resources/arquillian.xml:
```diff
@@ -0,0 +1,35 @@
+        <!--<property name="jbossHome">/jbossHome/path</property>-->
+        <!--</configuration> -->
+    </container>
+</arquillian>
```
Please see the ktchensink/src/test/resource/arquillian.xml file for the new container configuration section. It should look like this:

```
<!-- Example configuration for a remote JBoss Enterprise Application Platform 6 or AS 7 instance -->
<container qualifier="jboss" default="true">
    <!-- By default, arquillian will use the JBOSS_HOME environment variable.  Alternatively, the configuration below can be uncommented. -->
    <!--<configuration> -->
    <!--<property name="jbossHome">/path/to/jboss/as</property> -->
    <!--</configuration> -->
</container>
```

sgilda commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,276 @@
+
+
+		<version.jboss.arquillian>1.2.0.Alpha1</version.jboss.arquillian>
+		<version.org.jboss.arquillian.drone>1.2.0.Alpha1</version.org.jboss.arquillian.drone>
```
@rafabene : 
- Do we use Alpha1 versions?
- Did you combine  all the arquillian property names into just one property?

sgilda commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,184 @@
+
+For more information, see [Use JBoss Developer Studio or Eclipse to Run the Quickstarts](../README.md#useeclipse)
+
+Investigate the Console Output
```
I would like to see this section ditrectly below the `Run the Arquillian Tests` section. When you run the command, you want to immediately understand the output. After that, you can talk about running tests from JDBS.

sgilda commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,184 @@
+The session uses a rule inside the classpath, but you can  load rules from an url or 
+from a repository (i.e. guvnor in BRMS). 
+
+Test is built by Arquillian Drone using selenium in a trasparent way.
```
s/selenium/Selenium
s/trasparent/transparent

sgilda commented:
I tested this and it works great! I made a few comments in the files. I had a few other general comments:
- I don't see any comments in any of the classes or methods.  Could you add comments where it's not clear? :-)
- We can't assume that developers will be familiar with Drools, BRMS, or DeltaSpike. I would like the README file to go into a little more detail about this example and maybe a description of some of the files and classes and what they do. 
  - For example:
    -You might want to point out the rule definition in src/main/resources/SimpleRule.drl and that it gives a 10% discount for items greater than 0.
  - Which files in the cdi/drools/extension are important and why?

marcobattaglia commented:
Thank you very much for your suggestion!
I'm starting to work on.

sgilda commented:
@marcobattaglia : Would you mind squashing the commits down to just 1 commit? :-)

sgilda commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,199 @@
+
+This project demonstrates a simple CDI Portable Extension and some of the SPI classes used to complete that task. This particular extension explores InjectionTarget, ProcessAnnotatedType, interceptor bindings from CDI, and AnnotatedTypeBuilder from DeltaSpike to demonstrate how you can augment JavaEE6 with custom extension. The presentation layer and business layer are well decoupled and the busi ...
+
+Test is built by Arquillian Drone using selenium in a trasparent way.
```
Should Selenium be capitalized?
typo: s/trasparent/transparent

sgilda commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,199 @@
+What is it?
+-----------
+
+This project demonstrates a simple CDI Portable Extension and some of the SPI classes used to complete that task. This particular extension explores InjectionTarget, ProcessAnnotatedType, interceptor bindings from CDI, and AnnotatedTypeBuilder from DeltaSpike to demonstrate how you can augment JavaEE6 with custom extension. The presentation layer and business layer are well decoupled and the busi ...
```
Now the paragraph is harder to read. Sorry about that. Maybe this could be broken down into a couple of paragraphs as follows?

This project demonstrates a simple CDI Portable Extension and some of the SPI classes used to complete that task. This particular extension explores `InjectionTarget`, `ProcessAnnotatedType`, interceptor bindings from CDI, and `AnnotatedTypeBuilder` from DeltaSpike to demonstrate how you can augment Java EE 6 with custom extensions. 

The presentation layer and business layer are well decoupled and the business logic is augmented by the Drools as the rule engine. In Drools' philosophy, the world is a whole of facts that interact each other by rules. 

Our little world, behind the JSF page, is made up of billable LineItems that are modified by economic rules defined in `resources/SimpleRule.drl`. When you add or remove a `LineItem`, the rule engine triggers an action. To obtain this behavior, our managed beans are promoted to be facts using an amazing feature of CDI: portable extension. The implementation can be found in `DroolsExtension.java` and `LineItemInjectionTarget.java`. The rules are defined inside a knowledge session whose scope is managed by a producer. In this case, `ksession` lives inside a `ConversationScope`. 

You can also load rules from an URL or from a repository or guvnor in BRMS. Your repository then acts as a single point of truth and manages the logic of your application's world. From a business point of view, this means you can change economic rules without changing application code.

marcobattaglia replied:
ok I've committed. Sorry I don't understand when you ask to squashing the commit down to 1. I use git commit -a and then git push. Could you give me a tip about how to do ? Thanx

sgilda commented:
@marcobattaglia : This should help: http://stackoverflow.com/questions/5189560/how-can-i-squash-my-last-x-commits-together-using-git

My nick on IRC is sgilda. If you need help, ping me on freenode channel #jboss-jdf if you run into a problem or need help.

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/DroolsExtension.java:
```diff
@@ -0,0 +1,93 @@
+			.getLogger(DroolsExtension.class.getName());
+	
+/*
+ * Managed bean lifecycle is managed by weld that raise 
```
s/weld/cdi/

pmuir commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,205 @@
+Level: Intermediate
+Technologies: CDI, DeltaSpike, Drools, Arquillian Drone
+Summary: Creating a basic CDI extension with a custom @FactModel annotation that interacts with Rule Engine 
+Target Product: WFK
```
The target product for this needs to be BRMS.

pmuir commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,276 @@
+		<maven.compiler.target>1.6</maven.compiler.target>
+		<maven.compiler.source>1.6</maven.compiler.source>
+
+		<version.jboss.drools>5.5.0.Final</version.jboss.drools>
```
this needs moving to a BOM

marcobattaglia replied:
Sorry I don't know how to create a BOM on a public maven repository.

sgilda replied:
@marcobattaglia : let me know if the instructions below are not sufficient. :-)

marcobattaglia replied:
Hi S Gilda
I've prepared a BOM: https://github.com/marcobattaglia/jboss-bom/tree/jboss-javaee-6.0-with-drools/jboss-javaee-6.0-with-drools-expert
In this POM we have three dependencies to drools expert. If you consider that it's ok I'll pull a request. 
Thank you.

sgilda replied:
I am not a BOM exper, so I'll have to defer to @pmuir or @rafabene.

The only thing I question is why you named it `-with-drools-expert` rather than `-with-drools`. Does it need the `-expert` suffix?

Go ahead and issue the pull so someone can review it for you. :-)

marcobattaglia replied:
-expert suffix because Drools project consists out of several projects. I have linked only the dependencies related to Drools Expert that is the core of rule engine. Request just pulled.

pmuir commented on cdi-portable-extension-drools/pom.xml:
```diff
@@ -0,0 +1,276 @@
+		<version.jboss.drools>5.5.0.Final</version.jboss.drools>
+
+		<version.jboss.arquillian>1.1.1.Final</version.jboss.arquillian>
+		<version.org.jboss.arquillian.drone>1.1.1.Final</version.org.jboss.arquillian.drone>
```
This needs moving to a BOM.

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/DroolsExtension.java:
```diff
@@ -0,0 +1,93 @@
+ * - ProcessInjectionTarget raises when a managed bean is injected
+ * - ProcessAnnotatedType   raises before the container builds its metamodel
+ */
+	void beforeBeanDiscovery(@Observes BeforeBeanDiscovery bbd, BeanManager bm) {
```
This seems a bit redundant?

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/FactMethod.java:
```diff
@@ -0,0 +1,38 @@
+
+@Interceptor
+@FactMethodBinding
+public class FactMethod {
```
The whitespace here is unusual - you probably want to run the formatter over the whole thing

pmuir replied:
Also, we conventionally name an interceptor something like FactMethodInterceptor

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/LineItemInjectionTarget.java:
```diff
@@ -0,0 +1,90 @@
+ /*
+  * During injection this method adds the FactModel to ksession.
+  * In this way the managed bean become a fact and the rule engine 
+  * can apply rules to it.  * 
```
trailing *

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/LineItemInjectionTarget.java:
```diff
@@ -0,0 +1,90 @@
+  * @see javax.enterprise.inject.spi.InjectionTarget#inject(java.lang.Object, javax.enterprise.context.spi.CreationalContext)
+  */
+	@Override
+	public void inject(X instance, CreationalContext<X> ctx) {
```
Shouldn't this be done in produce()?

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/RuleSessionManager.java:
```diff
@@ -0,0 +1,104 @@
+import org.drools.io.ResourceFactory;
+import org.drools.runtime.StatefulKnowledgeSession;
+
+public class RuleSessionManager implements Serializable {
```
A comment on this class would be nice.

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/RuleSessionManager.java:
```diff
@@ -0,0 +1,104 @@
+			.getName());
+	private static final long serialVersionUID = 1L;
+	KnowledgeBase kbase;
+	StatefulKnowledgeSession ksession;
```
No private modifier here? Is that intentional?

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/RuleSessionManager.java:
```diff
@@ -0,0 +1,104 @@
+
+	private final static Logger log = Logger.getLogger(RuleSessionManager.class
+			.getName());
+	private static final long serialVersionUID = 1L;
```
Either use a generated serialVersionUID or omit it and suppress the warning.

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/cdi/drools/extension/RuleSessionManager.java:
```diff
@@ -0,0 +1,104 @@
+
+	}
+
+}
```
This whole class needs comments :-)

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/model/Manager.java:
```diff
@@ -0,0 +1,65 @@
+import org.jboss.as.quickstarts.cdi.drools.extension.FactModel;
+
+@Model
+public class Manager {
```
Class level comments?

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/model/Manager.java:
```diff
@@ -0,0 +1,65 @@
+	OrderBuilder ob;
+
+	@Inject
+	BeanManager beanManager;
```
private modifiers?

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/model/Manager.java:
```diff
@@ -0,0 +1,65 @@
+	 */
+	@Inject
+	@FactModel
+	LineItem lineItem;
```
Why not inject this into a initializer method and do away with the post construct?

pmuir commented on cdi-portable-extension-drools/src/main/java/org/jboss/as/quickstarts/model/OrderBuilder.java:
```diff
@@ -0,0 +1,68 @@
+
+@ConversationScoped
+@Stateful
+public class OrderBuilder {
```
Class level comments

pmuir commented:
I want to review again once updates are in.

sgilda commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,205 @@
+
+The presentation layer and business layer are well decoupled and the business logic is augmented by the Drools as the rule engine. In Drools' philosophy, the world is a whole of facts that interact each other by rules.
+
+Our little world, behind the JSF page, is made up of billable LineItems that are modified by economic rules defined in resources/SimpleRule.drl. When you add or remove a LineItem, the rule engine triggers an action taht assigns a discount. To obtain this behavior, our managed beans are promoted to be facts using an amazing feature of CDI: portable extension. The implementation can be found in Dro ...
```
I'm having trouble reading this sentence, particularly "behind the JSF page, is made up of..."

Maybe: The JSF page consists of billable LineItems that are modified by economic rules defined in resources/SimpleRule.drl.

There's also a typo in this sentence: When you add or remove a LineItem, the rule engine triggers an action taht assigns a discount.
s/taht/that

sgilda replied:
I'm still seeing "taht". Also, it might be easier to read if you put tick marks around file names and classes, for example: `resources/SimpleRule.drl`
`DroolsExtension.java`
`LineItemInjectionTarget.java`
`ConversationScope`

sgilda commented:
Great detail in "What is it?"!! I wish they were all this detailed!

sgilda commented:
Tested and worked as advertised!

sgilda commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,205 @@
+
+        mvn clean test -Parq-jbossas-remote
+
+This test will open the browser and it adds and removes some products (LineItems). In every step it checks the discount value that is setted by the rule engine.
```
You might want to emphasize this a bit because I missed it in the instructions and was surprised when it popped up. :-)

_Note: This test will open the browser and it adds and removes some products (LineItems). In every step it checks the discount value that is setted by the rule engine._

sgilda commented:
@marcobattaglia :
- To create a BOM, I believe you need to fork https://github.com/jboss-jdf/jboss-bom, clone it, create the BOM and issue a pull request. If you can't figure it out by looking at the existing BOM examples, I can try to find your some help with it.
- You can find information about the formatter file here: http://www.jboss.org/jdf/quickstarts/get-involved/. Click the "Don't forget to read the Contributing Guide" link and look under "General Guidelines". It's also  That information is also in the CONTRIBUTING.md file in the root of the quickstart directory. 

Let me know if you need anything else.

sgilda commented:
@marcobattaglia : Do you need any help with the BOM or formatting?

marcobattaglia commented:
Hi sgilda, sorry I was very busy. On 27 April my second daughter was born ;-)
Now I've committed BOM, but I don't know how to test it in my project. How can maven download the BOM if it isn't on a maven repo? 
About formatter there isn't nothing on the link above, only the commented header of java files about RedHat ecc. Do you use a particular formatter file in JBDS or can I use the default java formatter?

sgilda commented:
@marcobattaglia : Congratulations on the birth of your 2nd daughter! I'm sure you have your hands full right now. :-)

I posted information about the formatter in the comment above, but I wasn't very clear. The formatting information (plus additional quickstart guidelines) can be found in the CONTRIBUTING.md file at the root of the quickstart directory: 

```
    The sample project should be formatted using the JBoss AS profiles found at http://github.com/jboss/ide-config/tree/master/
```

@rafabene : Can you answer the question about how to test a BOM commit? I can't recall how to do that.

sgilda commented:
@marcobattaglia : Just checking in. :-)
Where do you stand with this quickstart? 
Were you able to test your BOM and issue a pull?
Have you finished the other changes?

marcobattaglia commented:
@sgilda 
@rafabene 
Hi, I've problem with BOM. 
The list of operations I've executed:
- Added   <version.jboss.drools>5.5.0.Final</version.jboss.drools>  in properties sections of parent pom
- Added   module  <module>jboss-javaee-6.0-with-drools</module> in parent pom
- I've installed in my local repo using mvn install
- I've added the dependency of BOM in my project, like others BOM for example deltaspike
- Here JBDS shows an error related to version of drools' bom, so I've changed from 1.0.4.Final to 1.0.7-SNAPSHOTS as it is in BOM project ...ok
- I've commented out drools dependencies, because they are included by bom,  but another error is shown: compile error "package org.drools does not exist"
- So I tried to comment out also deltaspike-core-api because it should be included by jboss-javaee-6.0-with-deltaspike, but there is an error also for deltaspike package.
  In conclusion the installed BOM is seen by my project (it checks correct version number) but libraries are not included in war.
  If you have any suggestion...

sgilda commented:
@marcobattaglia : I'll take a look. I probably won't be able to figure it out, but I'll try to find someone who can. :-)

Also, I ran the QS Tools utility against this quickstart and it reports a number of formatting violations. There are quite a few tab characters and incorrect indentations. Could you take a look at the CONTRIBUTING page (https://github.com/jboss-jdf/jboss-as-quickstart/blob/master/CONTRIBUTING.md) and make sure you formatted your project using the JBoss AS profiles found at http://github.com/jboss/ide-config/tree/master/. That page will also tell you how to run the QS Tools report against your quickstart to catch violations.

sgilda commented:
@marcobattaglia : This worked for me in command mode! I fetched the upstream quickstarts and rebased against upstream/master. Then I ran the "mvn clean  test -Parq-jbossas-remote" and saw the results described in the README file.

I am going to try in JDBS next.

In the meantime, it would be great if you could fix the formatting issues. :-)

sgilda commented:
@marcobattaglia : Your quickstart Maven test runs in JDBS exactly the same way all the other quickstarts, for example kitchensink and bean-validation do. I think this one is ready to go when you fix the formatting issues and typos.

sgilda commented on cdi-portable-extension-drools/README.md:
```diff
@@ -0,0 +1,205 @@
+
+Maven prints summary of performed tests into the console:
+
+-------------------------------------------------------
```
If you indent the following test results 4 spaces, it will render better in HTML.

marcobattaglia commented:
@sgilda I've committed after reformatting and cleaning up all project. This project is related to jboss-javaee-6.0-with-drools project under http://github.com:jboss-jdf/jboss-bom. It's important that both projects will be available together. In the pom.xml file (the parent) of all jboss-javaee-6.0-with-XXX projects I have added version.org.jboss.drools property and everything needed to include the drool's bom (as suggested by Rafael). Please check about versions of BOM projects. The general version.jboss.bom used in my quickstart project is the 1.0.4.Final, but the version of jboss-javaee-6.0-with-drools is 1.0.7-SNAPSHOT. I think that something has to be added in maven repo. 
I'm available for any questions. bye

sgilda commented:
@marcobattaglia : Thanks for the update. I think I'll wait for @rafabene to comment on the BOM updates when he gets back from vacation.

rafabene commented:
Just FIY the -with-drools BOM was merged to our repo!

marcobattaglia commented:
@rafabene Thank you !

sgilda commented:
@rafabene: Thanks!

@marobattaglia: Would you mind squashing the commits? :-)

rafabene commented:
JBoss BOM 1.0.7.CR2 released with -with-drools BOM

sgilda commented:
@rafabene : I just tried this. 
First, I got this error:
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]  
[ERROR]   The project org.jboss.as.quickstarts:jboss-as-cdi-add-interceptor-binding:7.1.2-SNAPSHOT (/home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/pom.xml) has 4 errors
[ERROR]     Non-resolvable import POM: Could not find artifact org.jboss.bom:jboss-javaee-6.0-with-drools:pom:1.0.4.Final in central (http://repo.maven.apache.org/maven2) @ line 86, column 17 -> [Help 2]
[ERROR]     'dependencies.dependency.version' for org.drools:knowledge-api:jar is missing. @ line 146, column 16
[ERROR]     'dependencies.dependency.version' for org.drools:drools-core:jar is missing. @ line 150, column 15
[ERROR]     'dependencies.dependency.version' for org.drools:drools-compiler:jar is missing. @ line 154, column 15

Baed upon your comment above, I modified the pom file as follows:  <version.jboss.bom>1.0.7.CR2</version.jboss.bom>

Then I get this error: 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]  
[ERROR]   The project org.jboss.as.quickstarts:jboss-as-cdi-add-interceptor-binding:7.1.2-SNAPSHOT (/home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/pom.xml) has 4 errors
[ERROR]     Non-resolvable import POM: Could not find artifact org.jboss.bom:jboss-javaee-6.0-with-drools:pom:1.0.7.CR2 in central (http://repo.maven.apache.org/maven2) @ line 86, column 17 -> [Help 2]
[ERROR]     'dependencies.dependency.version' for org.drools:knowledge-api:jar is missing. @ line 146, column 16
[ERROR]     'dependencies.dependency.version' for org.drools:drools-core:jar is missing. @ line 150, column 15
[ERROR]     'dependencies.dependency.version' for org.drools:drools-compiler:jar is missing. @ line 154, column 15

Am I modifying the wrong thing?

rafabene commented:
Sorry!

The new nexus version and the release script become incompatible and the artifacts were not publishing and reporting errors. That's why I'm working on https://issues.jboss.org/browse/JDF-364 - You can use 1.0.7.CR4 which is available on Maven Central.

marcobattaglia commented:
@sgilda 
Hi, do you need any helps about it? Do I have to change something in the code?
Please tell me.
bye

sgilda commented:
@marcobattaglia : I will try the 1.0.7.CR5 version of the JBoss BOM and let you know.

sgilda commented:
@marcobattaglia , @rafabene : I updated the pom.xml file to use `<version.jboss.bom>1.0.7.CR5</version.jboss.bom>` and tested it again.

When I run the tests, I get 7 compilation errors:

```
    [INFO] 7 errors 
    [INFO] -------------------------------------------------------------
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time: 4.541s
    [INFO] Finished at: Fri Jun 21 13:32:28 EDT 2013
    [INFO] Final Memory: 37M/325M
    [INFO] ------------------------------------------------------------------------
    [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.3.1:testCompile (default-testCompile) on project jboss-as-cdi-add-interceptor-binding: Compilation failure: Compilation failure:
    [ERROR] /home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:[17,40] error: cannot find symbol
    [ERROR] 
    [ERROR] package org.jboss.shrinkwrap.resolver.api
    [ERROR] /home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:[18,46] error: cannot find symbol
    [ERROR] 
    [ERROR] package org.jboss.shrinkwrap.resolver.api.maven
    [ERROR] /home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:[17,40] error: cannot find symbol
    [ERROR] 
    [ERROR] package org.jboss.shrinkwrap.resolver.api
    [ERROR] /home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:[18,46] error: cannot find symbol
    [ERROR] 
    [ERROR] package org.jboss.shrinkwrap.resolver.api.maven
    [ERROR] /home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:[29,8] error: cannot find symbol
    [ERROR] 
    [ERROR] class CdiExtensionDroolsTest
    [ERROR] /home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:[30,12] error: cannot find symbol
    [ERROR] 
    [ERROR] class CdiExtensionDroolsTest
    [ERROR] /home/sgilda/GitRepos/quickstart-jdf/cdi-portable-extension-drools/src/test/java/org/jboss/as/quickstarts/jboss_as_cdi_portable_extension_drools/CdiExtensionDroolsTest.java:[29,43] error: cannot find symbol
```

sgilda commented:
@marcobattaglia : 
First, in the `pom.xml` file, you need to update the `<version.jboss.bom>` version to 1.0.7.CR5.

@LightGuard helped me with the compile issues. It appears the shrinkwrap resolver API as changed as well. He was able to figure out `DependencyResolvers` was changed to `Resolvers`. So when I changed the import in CdiExtensionDroolsTest.java from `import org.jboss.shrinkwrap.resolver.api.DependencyResolvers;` to `import org.jboss.shrinkwrap.resolver.api.Resolvers;`, it went from 7 compile errors down to 5 compile errors.

It appears the `import org.jboss.shrinkwrap.resolver.api.maven.MavenDependencyResolver;` class is no longer in the package either. Jason searched and found a  helpful link here: 
 https://github.com/shrinkwrap/resolver/blob/2.0.0-beta-5/impl-maven/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/integration/PomDependenciesUnitTestCase.java

Jason also found these links:
- https://community.jboss.org/thread/215047?tstart=0
- https://community.jboss.org/thread/228683?tstart=0

He suggested you might want to talk to Karel if you have any questions.

Hope this helps!

LightGuard commented:
You may also find the README at https://github.com/shrinkwrap/resolver/tree/master helpful.

sgilda commented:
@marcobattaglia : Have you had a chance to look at this yet?

sgilda commented:
Sent email 8/2/2013.

marcobattaglia commented:
@sgilda Next week, yes. Sorry I was very busy implementing my first extension for JBoss EAP 6.

sgilda commented:
@marcobattaglia : Understood. Just wanted to get a status update. No pressure :-)

marcobattaglia committed:
Update existing Shrinkwrap examples to use 2.0.0 (BOM 1.0.7.CR5)

marcobattaglia committed:
Bug fix to use correctly ShrinkWrap 2.0.0 and to give correct artifact name

sgilda commented:
@marcobattaglia : I just tested this. It compile fine now, but the tests fail. The log is below. Any idea what is wrong?

---

##  T E S T S

Running org.jboss.as.quickstarts.jboss_as_cdi_portable_extension_drools.CdiExtensionDroolsTest
16:26:24.205 INFO - Command request: getNewBrowserSession[*firefox, http://localhost:8080, ] on session null
16:26:24.211 INFO - creating new remote session
16:26:24.325 WARN - Caution: '/usr/bin/firefox': file is a script file, not a real executable.  The browser environment is no longer fully under RC control
16:26:24.329 INFO - Allocated session dc9710470ad6472fa35ab990f1cc4c96 for http://localhost:8080, launching...
jar:file:/home/sgilda/.m2/repository/org/seleniumhq/selenium/selenium-server/2.29.0/selenium-server-2.29.0.jar!/customProfileDirCUSTFFCHROME
16:26:24.366 INFO - Preparing Firefox profile...
16:26:30.429 INFO - Launching Firefox...
16:26:33.610 INFO - Got result: OK,dc9710470ad6472fa35ab990f1cc4c96 on session dc9710470ad6472fa35ab990f1cc4c96
16:26:33.619 INFO - Command request: setSpeed[0, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:33.620 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
16:26:33.624 INFO - Command request: setTimeout[60000, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:33.664 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
16:26:33.735 INFO - Command request: open[http://127.0.0.1:8080/test/index.xhtml, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:33.735 WARN - you appear to be changing domains from http://localhost:8080 to http://127.0.0.1:8080/test/index.xhtml
this may lead to a 'Permission denied' from the browser (unless it is running as *iehta or *chrome,
or alternatively the selenium server is running in proxy injection mode)
16:26:35.557 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.560 INFO - Command request: type[id=form:txtDescription, demo] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.574 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.577 INFO - Command request: type[id=form:txtPrice, 24] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.592 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.595 INFO - Command request: click[id=form:btnBuy, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.633 INFO - Got result: ERROR: Command execution failure. Please search the user group at https://groups.google.com/forum/#!forum/selenium-users for error details from the log window.  The error message is: Value does not implement interface Event. on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.669 INFO - Command request: type[id=form:txtPrice, 100] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.681 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.686 INFO - Command request: click[id=form:btnBuy, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.709 INFO - Got result: ERROR: Command execution failure. Please search the user group at https://groups.google.com/forum/#!forum/selenium-users for error details from the log window.  The error message is: Value does not implement interface Event. on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.738 INFO - Command request: type[id=form:txtPrice, 100] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.749 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.754 INFO - Command request: click[id=form:btnBuy, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.770 INFO - Got result: ERROR: Command execution failure. Please search the user group at https://groups.google.com/forum/#!forum/selenium-users for error details from the log window.  The error message is: Value does not implement interface Event. on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.823 INFO - Command request: click[xpath=//tr[last()]/td/input, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:35.833 INFO - Got result: ERROR: Command execution failure. Please search the user group at https://groups.google.com/forum/#!forum/selenium-users for error details from the log window.  The error message is: Value does not implement interface Event. on session dc9710470ad6472fa35ab990f1cc4c96
16:26:36.020 INFO - Command request: close[, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:36.114 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
16:26:36.118 INFO - Command request: testComplete[, ] on session dc9710470ad6472fa35ab990f1cc4c96
16:26:36.118 INFO - Killing Firefox...
16:26:36.176 INFO - Got result: OK on session dc9710470ad6472fa35ab990f1cc4c96
Tests run: 4, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 20.838 sec <<< FAILURE!

Results :

Tests in error: 
  assertFilledUnderOneHundred(org.jboss.as.quickstarts.jboss_as_cdi_portable_extension_drools.CdiExtensionDroolsTest): ERROR: Command execution failure. Please search the user group at https://groups.google.com/forum/#!forum/selenium-users for error details from the log window.  The error message is: Value does not implement interface Event.
  assertFilledOverOneHundred(org.jboss.as.quickstarts.jboss_as_cdi_portable_extension_drools.CdiExtensionDroolsTest): ERROR: Command execution failure. Please search the user group at https://groups.google.com/forum/#!forum/selenium-users for error details from the log window.  The error message is: Value does not implement interface Event.
  assertFilledOverTwoHundred(org.jboss.as.quickstarts.jboss_as_cdi_portable_extension_drools.CdiExtensionDroolsTest): ERROR: Command execution failure. Please search the user group at https://groups.google.com/forum/#!forum/selenium-users for error details from the log window.  The error message is: Value does not implement interface Event.
  assertFilledNewlyUnderOneHundred(org.jboss.as.quickstarts.jboss_as_cdi_portable_extension_drools.CdiExtensionDroolsTest): ERROR: Command execution failure. Please search the user group at https://groups.google.com/forum/#!forum/selenium-users for error details from the log window.  The error message is: Value does not implement interface Event.

Tests run: 4, Failures: 0, Errors: 4, Skipped: 0

marcobattaglia commented:
Ahiaiaii. On my fedora it works fine, so that I think it's a problem of timing and/or browser/selenium.  I will do a check
thank you

marcobattaglia commented:
@sgilda 
I've found this:
http://code.google.com/p/selenium/issues/detail?id=5554

it seems to be an issue related to selenium (the client tester) with the new version of firefox (Aurora). The click event is not recognized. The selenium library 2.34 solves this problem. Our selenium library is 2.29 and it is loaded by :
               &lt;dependency&gt;
            &lt;groupId&gt;org.jboss.arquillian.extension&lt;/groupId&gt;
            &lt;artifactId&gt;arquillian-drone-selenium-depchain&lt;/artifactId&gt;
            &lt;type&gt;pom&lt;/type&gt;
            &lt;scope&gt;test&lt;/scope&gt;
        &lt;/dependency&gt;
I've also tried to do an upgrade to WebDrive but without good result. In this case my firefox (version 20.0) alerts plugin not supported.
What is your firefox version?
If it is Aurora could you try to change pom in this way:
&lt;version.org.jboss.arquillian.drone&gt;1.2.0.Alpha2&lt;/version.org.jboss.arquillian.drone&gt;
instead 
&lt;version.org.jboss.arquillian.drone&gt;1.1.1.Final&lt;version.org.jboss.arquillian.drone&gt;
before lauching test?
Thank you

sgilda commented:
@marcobattaglia : I'm using Firefox 22.0 on Fedora 19. It appears that Firefox 22 is not supported by Selenium RC. 

I also found these links that talk about my error: 'Value does not implement interface Event.'
- https://groups.google.com/forum/#!searchin/selenium-users/value$20does$20not$20implement$20interface$20event%7Csort:relevance/selenium-users/y1m9xX68RgU/4wiEhcn-L5oJ
- https://groups.google.com/forum/#!topic/selenium-users/1V5KktYYHI0

Does this quickstart require Firefox? Chrome is my default browser. 
What happens if a developer doesn't have Firefox installed?

marcobattaglia commented:
src/test/resources/arquillian.xml --> tag browserCapabilities. ... You can choose another browser to test.

sgilda commented:
I changed that file to the following:

```
    <extension qualifier="webdriver">
       <property name="browserCapabilities">chrome</property>
    </extension>
```

Yet, I still see that it opens Firefox and I get the same error:

08:18:34.959 INFO - Command request: getNewBrowserSession[*firefox, http://localhost:8080, ] on session null
08:18:34.980 INFO - creating new remote session
08:18:35.040 WARN - Caution: '/usr/bin/firefox': file is a script file, not a real executable.  The browser environment is no longer fully under RC control
08:18:35.044 INFO - Allocated session 9d63a5e93cf7478aa35d2046f29ea3d5 for http://localhost:8080, launching...
jar:file:/home/sgilda/.m2/repository/org/seleniumhq/selenium/selenium-server/2.29.0/selenium-server-2.29.0.jar!/customProfileDirCUSTFFCHROME
08:18:35.088 INFO - Preparing Firefox profile...
08:18:36.944 INFO - Launching Firefox...

I assume I did not change it correctly. Still investigating. 

UPDATE: Found this doc: https://docs.jboss.org/author/display/ARQ/Drone
Under "Selenium 2 (WebDriver) configuration", I see "chrome" in the list of valid values, so this should be correct and should have worked.

sgilda commented:
Found an example showing '*googlechrome' on this website: http://arquillian.org/guides/functional_testing_using_drone/

Tried that and I still see it trying to open Firefox:
08:29:22.384 INFO - Command request: getNewBrowserSession[*firefox, http://localhost:8080, ] on session null
08:29:22.393 INFO - creating new remote session
08:29:22.463 WARN - Caution: '/usr/bin/firefox': file is a script file, not a real executable.  The browser environment is no longer fully under RC control
08:29:22.466 INFO - Allocated session c5d9783a3a3049f8a2a90668d540f160 for http://localhost:8080, launching...
jar:file:/home/sgilda/.m2/repository/org/seleniumhq/selenium/selenium-server/2.29.0/selenium-server-2.29.0.jar!/customProfileDirCUSTFFCHROME
08:29:22.503 INFO - Preparing Firefox profile...
08:29:23.709 INFO - Launching Firefox...

I can see that it generates the correct XML in the target directory:
target/test-classes/arquillian.xml:   <property name="browserCapabilities">*googlechrome</property>

Also ried it with and without the leading *.
I am testing against JBoss AS 7.1.1.

marcobattaglia commented:
@sgilda
I think we have two problems: 
1) it doesn't work with the new version of firefox: in particular this  browser.click("id=form:btnBuy"); doesn't work because there is an issue and selenium libraries must be updated to 2.34.0 version 
2) We cannot launch chrome because browserCapabilities tag seems to not work or something is wrong in the configuration

In more:
3) there are a lot of new features related to arquillian drone ( https://gist.github.com/kpiwko/1848342 ). I think that the more interesting is about executing multiple drones (firefox and chrome for example).

We can work to add the new features of arquillian drone in this quickstart resolving 1) and 2) .
I need only to know if it is possible without breaking dependencies version, because kpiwko in this link (https://gist.github.com/kpiwko/1848342) suggests to remove Java EE 6 Spec BOM file with a Java EE 6 + Tools BOM from pom.xml

Before starting please tell me what do you think about it.
The alternative is to resolve points 1 and 2, without adding point 3.
Thank you

sgilda commented:
@pmuir, @rafabene : This is way over my head. Can either of you respond?

rafabene commented:
I think this gist https://gist.github.com/kpiwko/1848342 is a little bit out of date. I made some updates on -with-tools BOM to permit it work with ShrinkWrap resolver: https://github.com/jboss-jdf/jboss-bom/commit/a2fb096967d301e2731b50a9fb070b6fa79b1fd2 I think this solved also the problem with Drone too 

It worth try using Drone + -with-tools BOM. If it didn't work we should fix the BOM.

sgilda commented:
@marcobattaglia : We will be closing all outstanding quickstart pull requests this week so we can reorganize the quickstarts. Do you think you'll be able to finish this one up in the next day or two? I'm still not clear how to get it to use Chrome instead of Firefox.

I think the new Arquillian features (your issue 3.) could be added down the road, don't you? I think there are plans to enhance many of the existing quickstarts with regards to Arquillian tests.

marcobattaglia commented:
@sgilda @rafabene  :
ok. we can remove issue3.

Issue1: It can be resolved by updating selenium version to 2.34 (both selenium server library and selenium) Now  the arquillian-drone-bom is using version 2.29.This version has got a bug with the new version of firefox: it cannot simulate a click in the browser!

Issue2: It's correct that only firefox browser works, because the DefaultSelenium Drone has got only firefox as BrowserCapabilities. In order to have other browser the Drone must be changed to WebDriver. It has actually some issue with Firefox (http://stackoverflow.com/questions/15814695/selenium-webdriver-does-not-connect-with-firefox-20) and another upgrade will be required.

Conclusion:
If you want it's possible to resolve issue1 by changing arquillian-drone-bom in order to use the new version of selenium, or we can to change test in normal test without using a browser Drone.

sgilda commented:
@marcobattaglia : We need to resolve issue #1 before merging. What would need to be done to upgrade to Selenium 2.34? Is this a BOM file change or something else?

marcobattaglia commented:
@sgilda 
In  jboss-javaee-6.0-with-tools-1.0.7-SNAPSHOT.pom

is imported as dependency org.jboss.arquillian.extension

&lt;dependency&gt;
&lt;groupId&gt;org.jboss.arquillian.extension&lt;/groupId&gt;
&lt;artifactId&gt;arquillian-drone-selenium-depchain&lt;/artifactId&gt;
&lt;version&gt;${version.org.jboss.arquillian.extension.drone}&lt;/version&gt;
&lt;type&gt;pom&lt;/type&gt;
&lt;/dependency&gt;

this imports:
selenium-api version: 2.29.0
selenium-java  version: 2.29.0
selenium-server version: 2.29.0
selenium-remote-driver version: 2.29.0
selenium-support version: 2.29.0
selenium-safari-driver version: 2.29.0

These versions have to be upgraded to 2.34
I have tested the error on version 22 on firefox and there is no workaround, all events simulated on browser by selenium version 2.29 don't work.

I think the bug on click event is a serious bug and it justify an upgrade of arquillian bom.

Thank you for all your helps during these months

Bye

pmuir commented:
As this is a totally new quick start, it should be easy to apply against the new repo structure as well, with a simple mv :-)

sgilda commented:
@pmuir: I'm not clear what you mean. 
Should we update the BOM?
Should we wait to merge this after the reorg? 
Should we  ??? 
:-)

pmuir commented:
@sgilda, I mean it would be easy to merge this one after the reorg, as it's a new quickstart :-)

marcobattaglia commented:
it's ok for me. I can work also on new arquillian drones in order to have multiple browsers to test.

sgilda commented:
@marcobattaglia : Sounds good. I'll close the pull for now and you can submit it after the reorg is complete. Thanks!

:
